lib/src/expression/parser.dart

```diff
@@ -0,0 +1,128 @@
+import 'package:analyzer/analyzer.dart';
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/token.dart';
+import 'package:analyzer/dart/ast/standard_ast_factory.dart';
+import 'package:analyzer/error/listener.dart';
+import 'package:analyzer/file_system/memory_file_system.dart';
+import 'package:analyzer/src/dart/ast/token.dart';
+import 'package:analyzer/src/dart/scanner/reader.dart';
+import 'package:analyzer/src/dart/scanner/scanner.dart';
+import 'package:analyzer/src/generated/parser.dart';
+import 'package:analyzer/src/generated/source.dart';
+
+final _resourceProvider = new MemoryResourceProvider();
+
+class _ThrowingListener implements AnalysisErrorListener {
+  const _ThrowingListener();
+
+  @override
+  void onError(AnalysisError error) {
+    throw new FormatException(
+      error.toString(),
+      error.source.contents.data,
+      error.offset,
+    );
+  }
+}
+
+/// Parses a template [expression].
+Expression parseExpression(
+  String expression, {
+  bool deSugarPipes: true,
+  String sourceUrl: '/~test.dart',
+}) {
+  final source = _resourceProvider
+      .newFile(
+        sourceUrl,
+        expression,
+      )
+      .createSource();
+  final reader = new CharSequenceReader(expression);
+  final listener = const _ThrowingListener();
+  final scanner = new Scanner(source, reader, listener);
+  final parser = new _NgExpressionParser(
+    source,
+    listener,
+    deSugarPipes: deSugarPipes,
+  );
+  return parser.parseExpression(scanner.tokenize());
+}
+
+/// Extends the Dart language to understand the current Angular 'pipe' syntax.
+///
+/// Based on https://github.com/dart-lang/angular_analyzer_plugin/pull/160
+class _NgExpressionParser extends Parser {
+  final bool _deSugarPipes;
+
+  _NgExpressionParser(
+    Source source,
+    AnalysisErrorListener errorListener, {
+    bool deSugarPipes,
+  })
+      : _deSugarPipes = deSugarPipes,
+        super(source, errorListener);
+
+  @override
+  Expression parseBitwiseOrExpression() {
+    if (!_deSugarPipes) {
+      return super.parseBitwiseOrExpression();
+    }
+    Expression expression;
+    if (currentToken.keyword == Keyword.SUPER &&
+        currentToken.next.type == TokenType.BAR) {
+      expression = new SuperExpression(getAndAdvance());
+    } else {
+      expression = parseBitwiseXorExpression();
+    }
+    while (currentToken.type == TokenType.BAR) {
+      expression = new BinaryExpression(
```

Comment:
Rather than creating a BinaryExpression and breaking it apart later, I would just getAndAdvance(), parseBitwiseXorExpression() and pass "expression, pipeExpression" into the next function